crates/kitsune_p2p/dht/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+//! Defines the structure of the DHT, the objects which inhabit it, and the operations
+//! on these objects.
+//!
+//! ## Structure and contents of the DHT
+//!
+//! The DHT is populated by Agents and Ops. Agents introduce new Ops to the DHT,
+//! and are responsible for storing their own Ops and the Ops of other Agents.
+//!
+//! Each Agent has a fixed Location, and each Op has both a fixed
+//! Location and a fixed Timestamp.
+//!
+//! Locations are one-dimensional, and are best thought of as positions
+//! on a circle. They are represented by `Wrapping<u32>`.
+//! Timestamps are best thought of as positions on a timeline, and are represented
+//! by an i64 which denotes microseconds since the UNIX epoch.
+//! The Location dimension is circular, and the Time dimension is linear.
+//!
+//! Each Agent has an Arc which marks out a portion of the circle of Locations.
+//! Arcs have a fixed starting point, and extends "clockwise" by some length which
+//! is chosen by Kitsune and changes over time. Every Agent is responsible for
+//! holding a copy of any Ops whose location overlaps their Arc. As we will see
+//! later, Arcs are *quantized*.
+//!
+//! ### Geometrical interpretation
+//!
+//! This crate works heavily with the concept of a two-dimensional cylindrical
+//! "spacetime" surface -- all terminology in this crate is built with this
+//! model in mind, so it helps to explain it a bit here.
+//!
+//! If we think of Location as a spatial dimension,
+//! and Time as a temporal dimension, we can think of the DHT as a
+//! two-dimensional *spacetime*.
+//!
+//! The Location dimension is circular, i.e. the endpoints are connected,
+//! but the Time dimension is linear and constantly expanding. So, these two
+//! dimensions form a cylindrical surface, with its base at the origin time of
+//! the network, and a constantly increasing height as time marches forward.
+//!
+//! So:
+//! - An Op can be thought of a point somewhere on the surface of this cylinder,
+//! - An Agent can be thought of as a line perpendicular to the base, extending
+//!   across the entire length of the cylinder.
+//! - An Arcs can be thought of as a rectangular "stripe" made by sweeping the
```

Comment:
Yes, that diagram works. As long as the Op is actually on the side of the cylinder and not the top. I think what is written here corresponds to what you're saying, but maybe is confusing. Perhaps because I say "An Agent can be thought of as a line". It's not necessary to think of an Agent as such, but I think it makes sense to think of an Arc as originating from a vertical line and extending clockwise from that line in some direction. Hence "sweeping out" the area of the rectangle. Perhaps this image is a bit too dynamic though.